### PR TITLE
refactor(pubsub): add a tracing batch sink

### PR DESCRIPTION
**Pull request by alevenberg**
Part of https://github.com/googleapis/google-cloud-cpp/issues/13171

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13204)

### CODE REVIEW

File: google/cloud/pubsub/internal/tracing_batch_sink.cc [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  return batch_sink_spans;
+}
+
+}  // namespace
+
+/**
+ * Records spans related to a batch messages across calls and
+ * callbacks in the `BatchingPublisherConnection`.
+ */
+class TracingBatchSink : public BatchSink {
```
Comment (devbww) [inline google/cloud/pubsub/internal/tracing_batch_sink.cc:141]: Nit:  It looks like this could also go into the anonymous namespace.

Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_batch_sink.cc:139]: Done

File: google/cloud/pubsub/internal/tracing_batch_sink.cc [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    // such that each child span will have exactly batch_size elements or less.
+    batch_sink_spans.push_back(
+        MakeChild(batch_sink_parent, count++, MakeLinks(i, cut(i))));
+  }
+
+  return batch_sink_spans;
+}
+
+/**
+ * Records spans related to a batch messages across calls and
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/tracing_batch_sink.cc:136]: s/batch messages/batch of messages/

File: google/cloud/pubsub/internal/tracing_batch_sink_test.cc [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+        return make_ready_future(make_status_or(MakeResponse(request)));
+      });
+
+  auto batch_sink = MakeTestBatchSink(std::move(mock));
+
+  auto initial_spans = {message_span};
+  AddMessages(initial_spans, batch_sink);
+
+  auto response = batch_sink->AsyncPublish(MakeRequest(1)).get();
+  ASSERT_THAT(response, IsOk());
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/tracing_batch_sink_test.cc:166]: nit: If we don't care about what is in `response`, we should prefer `EXPECT_THAT(...)` so execution will continue if the expectation fails.

Here and below.

File: google/cloud/pubsub/internal/tracing_batch_sink_test.cc [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+
+  auto spans = span_catcher->GetSpans();
+  EXPECT_THAT(
+      spans,
+      Contains(AllOf(SpanHasInstrumentationScope(), SpanKindIsProducer(),
+                     SpanNamed("publish"),
+                     SpanHasAttributes(OTelAttribute<std::int64_t>(
+                         sc::kMessagingBatchMessageCount, kDefaultMaxLinks)),
+                     SpanLinksSizeIs(128))));
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/tracing_batch_sink_test.cc:312]: s/128/kDefaultMaxLinks/

File: google/cloud/pubsub/internal/tracing_batch_sink.cc [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+            internal::EndSpan(*span);
+          }
+          internal::DetachOTelContext(oc);
+          return f;
+        });
+  }
+
+  void ResumePublish(std::string const& ordering_key) override {
+    child_->ResumePublish(ordering_key);
+  };
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/tracing_batch_sink.cc:187]: nit: we should have a super simple unit test for `ResumePublish`.

File: google/cloud/pubsub/internal/tracing_batch_sink.cc [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        .then([oc = opentelemetry::context::RuntimeContext::GetCurrent(),
+               spans = std::move(batch_sink_spans),
+               message_spans = std::move(message_spans)](auto f) mutable {
+          for (auto& span : message_spans) {
+            span->AddEvent("gl-cpp.publish_end");
+          }
+          for (auto& span : spans) {
+            internal::EndSpan(*span);
+          }
+          internal::DetachOTelContext(oc);
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/tracing_batch_sink.cc:180]: I don't want to overcomplicate things, but we don't really have a test that would notice if this line was removed.

Consider in a follow up something like changing most tests to just:

```cc
  EXPECT_CALL(*mock, AsyncPublish(IsProtoEqual(MakeRequest(1))))
      .WillOnce(Return(ByMove(make_ready_future(make_status_or(MakeResponse(request))))));
  });
```

Then having a test dedicated to scope management that looks like:
```cc
TEST(TracingMessageBatch, Scope) {
  // ...
  EXPECT_CALL(*mock, AsyncPublish(IsProtoEqual(MakeRequest(1))))
      .WillOnce([](google::pubsub::v1::PublishRequest const& request) {
        EXPECT_TRUE(ThereIsAnActiveSpan());
        EXPECT_TRUE(OTelContextCaptured());
        return make_ready_future(make_status_or(MakeResponse(request)));
      });
  // ...
  auto response = batch_sink->AsyncPublish(MakeRequest(1)).then([] {
    EXPECT_FALSE(ThereIsAnActiveSpan());
    EXPECT_FALSE(OTelContextCaptured());
  }).get();
  // ...
}
```

**Review by dbolduc — APPROVED** [review]
This is in service of a refactor, but I am not sure I would call it a refactor.

It wasn't bad to review. I just compared the files to `tracing_message_batch*`.

Hypocritical[^1] Comment: I do get nervous when we add a new implementation with a new set of tests. That is slightly more error prone then refactoring an old implementation, and ensuring that its tests still pass. I am not sure I would have done anything differently though.

[^1]: Do not look at the `bigtable/internal/legacy_*` classes and tests

**Comment by alevenberg** [comment]
> This is in service of a refactor, but I am not sure I would call it a refactor.
> 
> It wasn't bad to review. I just compared the files to `tracing_message_batch*`.
> 
> Hypocritical[1](#user-content-fn-1-24cc6ce52a46f34b1c1bb1c250b3f5df) Comment: I do get nervous when we add a new implementation with a new set of tests. That is slightly more error prone then refactoring an old implementation, and ensuring that its tests still pass. I am not sure I would have done anything differently though.
> 
> ## Footnotes
> 1. Do not look at the `bigtable/internal/legacy_*` classes and tests [↩](#user-content-fnref-1-24cc6ce52a46f34b1c1bb1c250b3f5df)

Yeah, when moving files I used be able to branch them, but I don't know how to/if you can with git. This is the best I could do to break down the PR changes.

File: google/cloud/pubsub/internal/tracing_batch_sink_test.cc (hunk 4 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_batch_sink_test.cc:312]: Good catch. Done.

File: google/cloud/pubsub/internal/tracing_batch_sink.cc (hunk 2 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_batch_sink.cc:136]: Done

File: google/cloud/pubsub/internal/tracing_batch_sink.cc (hunk 5 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_batch_sink.cc:187]: Done

File: google/cloud/pubsub/internal/tracing_batch_sink.cc (hunk 6 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_batch_sink.cc:180]: I agree that is better than what I have now. Thanks! Fixed.

**Comment by alevenberg** [comment]
Thanks for the thorough review @dbolduc :)

**Pull request merged**